strawberryfields/backends/fockbackend/ops.py

```diff
@@ -500,3 +498,69 @@ def hermiteVals(q_mag, num_bins, m_omega_over_hbar, trunc):
         Hvals[i] = 2 * x * Hvals[i - 1] - 2 * (i - 1) * Hvals[i - 2]
 
     return q_tensor, Hvals
+
+
+def gkp_displacements(t, k, epsilon):
+    """
+    Helper function to generate the displacements parameters associated with the teeth of
+    GKP computational basis state k.
+    """
+    return np.sqrt(0.5 * np.pi) * (2 * t + k) / np.cosh(epsilon)
+
+
+def gkp_coeffs(t, k, epsilon):
+    """
+    Helper function to generate the coefficient parameters associated with the teeth of
+    GKP computational basis state k.
+    """
+    return np.exp(-0.5 * np.pi * np.tanh(epsilon) * (k + 2 * t) ** 2)
+
+
+@functools.lru_cache()
+def squaregkpBasisState(i, amplepsilon, amplcutoff, cutoff):
+    """
+    Generate the Fock expansion of a computational GKP basis state.
+
+    Args:
+        i (int): a computational basis state label, can be either 0 or 1
+        amplepsilon (float): finite energy parameter of the state
+        amplcutoff (float): this determines how many terms to keep in the Hilbert space expansion
+        cutoff (int): Fock space truncation
+
+    Returns
+        (array): the expansion of the ith computational basis state in the Fock basis
+
+    """
+    z_max = int(np.ceil(np.sqrt(-0.25 / np.pi * np.log(amplcutoff) / np.tanh(amplepsilon))))
+    coeffs = [gkp_coeffs(t, i, amplepsilon) for t in range(-z_max, z_max + 1)]
+    r = -0.5 * np.log(np.tanh(amplepsilon))
+    alphas = [gkp_displacements(t, i, amplepsilon) for t in range(-z_max, z_max + 1)]
+    num_kets = len(alphas)
+    ket = [coeffs[j] * displacedSqueezed(alphas[j], 0, r, 0, cutoff) for j in range(num_kets)]
+    ket /= np.linalg.norm(ket)
```

Comment:
Should normalization be applied at this stage or after multiplying the qubit basis states by their qubit coefficients? Eq. C24 in the paper is a subnormalized wavefunction after all.

E(eps) is a not a trace preserving operator, so:
E(eps)[a|0> + b|1>]/Norm(eps,a,b) 
is not necessarily the same as:
a[ E(eps)|0>/Norm(eps,1,0) ] + b[ E(eps)|1>/Norm(eps,0,1) ]

A simple test could be to check whether the Wigners match for some random qubit states for intermediate valued eps~0.2-1, where the normalization constants are more likely to noticeably differ. If we are only currently testing for the 0 state, they will always match.

A pathological case is the -1 Hadamard eigenstate which only has support on Fock states 4n+2, so in the limit eps->infty, it should go the the n=2 Fock state, whereas if you build it as a superposition of finite energy 0 and 1 states, it will go to n=0. GKPs are weird and awesome.

Reply:
Where do you think is better to apply the normalization?

Reply:
Equivalently, what is consistent with what is done in the bosonic backend?

Reply:
Based on what I understand the code to be doing, let's try normalizing at the very end of the state construction in line 565? Basically ket0 and ket1 in 562 and 563 will come subnormalized, then the whole state gets normalized once ket0 and ket1 are multiplied by the qubit coefficients. That would reflect how normalization is done in the paper. If that doesn't work, I think we'll have to dive deeper into what is going on.